### PR TITLE
Add OTEL tracing to the vault client

### PR DESCRIPTION
**Pull request by liam-j-bennett**
### Changes:
 - Add context propagation to the KVStore/Vault clients
 - Add the otel library as a dependecy (`go mod tidy`)
 - Start a new span within the lowest vault calls
 - Simplified error handling to allow for deferal of span closure

### Purpose
From the story:

> Create a wrapping library for all Vault calls that adds a span to an existing trace within services
> AC:
> Owner MUST check with Backplane and D&R teams to check for existing code
> Any library SHOULD be created/moved to a shared repository under the Backplane team
> Vault data MUST be namespaced within the span (e.g. vault.<data_name>)
> Span MUST contain the following Vault specific data:
> API route
> Request status
> Retries
> Metadata (i.e. any metadata from the request - e.g. Timestamp, id)
> Vault client version
> Vault wrapper version

Some of the data has not been added due to the vault SDK not providing access to the data, or not providing it in an easier way. These will be added to a separate ticket for adding at a later date.

### CODE REVIEW

**Comment by liam-j-bennett** [comment]
@RicardoLinck Please add anyone else you feel would be a good reviewer on this work

File: go.mod [hunk 1]
```diff
@@ -14,8 +14,8 @@ require (
 	github.com/pierrec/lz4 v2.0.5+incompatible // indirect
 	github.com/ryanuber/go-glob v1.0.0 // indirect
 	github.com/sirupsen/logrus v1.4.2
-	github.com/smartystreets/goconvey v1.6.4
-	github.com/stretchr/testify v1.3.0 // indirect
+	go.opentelemetry.io/otel v0.20.0
```
Comment (RicardoLinck) [inline go.mod]: In other projects we are using v1.0.0-RC1 -> https://github.com/magneticio/vamp-cloud-lamia/blob/master/go.mod#L37-L41

File: kvstore/vault.go [hunk 2]
```diff
@@ -152,15 +157,35 @@ func valueMap(value string) map[string]interface{} {
...
+	ctx, span := c.Tracer.Start(ctx, "vault request")
+	span.SetAttributes(attribute.Key("path").String(path), attribute.Key("action").String(action))
+	return ctx
+}
+
+func (c *VaultKeyValueStore) afterRequest(ctx context.Context, err error) {
+	span := trace.SpanFromContext(ctx)
+	defer span.End()
+	if err != nil {
+		span.SetAttributes(attribute.Key("error").String(err.Error()))
```
Comment (RicardoLinck) [inline kvstore/vault.go]: I think we are using `error` as a boolean, at least it seems to be the behaviour from gRPC instrumentation:
https://ui.honeycomb.io/circleci/datasets/distributed-traces/result/csFqoPCAAn

File: kvstore/vault.go [hunk 3]
```diff
@@ -152,15 +157,35 @@ func valueMap(value string) map[string]interface{} {
 
 // Vault API
 
-func (c *VaultKeyValueStore) GetData(key string, version int) (map[string]interface{}, error) {
+func (c *VaultKeyValueStore) beforeRequest(ctx context.Context, path string, action string) context.Context {
+	ctx, span := c.Tracer.Start(ctx, "vault request")
+	span.SetAttributes(attribute.Key("path").String(path), attribute.Key("action").String(action))
```
Comment (RicardoLinck) [inline kvstore/vault.go]: ~ Maybe renaming `action` to `method`? That's the pattern for HTTP instrumentation:
https://ui.honeycomb.io/circleci/datasets/distributed-traces/result/A1rDKpcxnij

**Review by robinsturmeit — APPROVED** [review]

**Review by RicardoLinck — APPROVED** [review]

**Review by dmitry-onishchenko — APPROVED** [review]

**Review by Salec — COMMENTED** [review]
lgtm

**Pull request merged**